core/src/main/scala/com/nvidia/spark/rapids/tool/analysis/AppSQLPlanAnalyzer.scala

```diff
@@ -319,15 +338,82 @@ class AppSQLPlanAnalyzer(app: AppBase, appIndex: Int) extends AppAnalysisBase(ap
         val med = Math.max(taskInfo.med, driverInfo.med)
         val total = Math.max(taskInfo.total, driverInfo.total)
 
-        Some(SQLAccumProfileResults(appIndex, metric.sqlID,
+        val sqlAccumProileResult = SQLAccumProfileResults(appIndex, metric.sqlID,
           metric.nodeID, metric.nodeName, metric.accumulatorId, metric.name,
-          min, med, max, total, metric.metricType, metric.stageIds.mkString(",")))
+          min, med, max, total, metric.metricType, metric.stageIds.mkString(","))
+
+        if (isIODiagnosticMetric(metric.name)) {
+          updateIODiagnosticMetrics(sqlAccumProileResult)
+        }
+
+        Some(sqlAccumProileResult)
       } else {
         None
       }
     }
   }
 
+  /**
+   * Generate IO-related diagnostic metrics for the SQL plan:
+   * output rows, scan time, output batches, buffer time, shuffle write time, fetch wait time, GPU
+   * decode time.
+   * @return a sequence of IODiagnosticResult
+   */
+  def generateIODiagnosticAccums(): Seq[IODiagnosticResult] = {
+    val zeroRecord = StatisticsMetrics.ZERO_RECORD
+    IODiagnosticMetrics.toSeq.flatMap { case ((sqlId, nodeId, nodeName, stageIds), sqlAccums) =>
+      stageIds.split(",").filter(_.nonEmpty).map(_.toInt).flatMap { stageId =>
+        val stageDiagnosticInfo = HashMap.empty[String, StatisticsMetrics]
+
+        sqlAccums.foreach { sqlAccum =>
+          val stageTaskIds = app.taskManager.getAllTasksStageAttempt(stageId).map(_.taskId).toSet
+          val accumInfo = app.accumManager.accumInfoMap.getOrElse(sqlAccum.accumulatorId,
+            new AccumInfo(AccumMetaRef(0L, AccumNameRef(""))))
+          val taskUpatesSubset =
```

Comment:
This blocks seems similar to calculating the statistics in `generateStageLevelAccums()`. Can we use a helper method to calculate statistics in both functions?

Reply:
thanks @parthosa, created helper functions `getStageTaskIds` and `getAccumInfoStatisticsInStage`